contents/java/streams-an-introduction.md

```diff
@@ -0,0 +1,75 @@
+# Java 8 Streams - An Introduction
+
+Authors: Lee Yi Min
+
+## Overview
+
+In Java 8, we were introduced to new features such as lambda expressions and streams. If you weren't familiar with the concept of functional programming, you might be (like me,) silently screaming in your head as you stare at a chunk of code infused with lambda expressions and stream operations.
+
+However, oftentimes, it's hard not to admire the conciseness of the codes utilising these features. For example, you might want to calculate the mean height of male students given a list of students. Using traditional for-loops, you may write something like this:
+
+```
+double totalMaleHeight = 0;
+int noOfMales = 0;
+for (Student student: students) {
+  if (student.isMale()) {
+    totalMaleHeight += student.getHeight();
+    noOfMales++;
+  }
+}
+double meanMaleHeight = totalMaleHeight / noOfMales;
+```
+
+However, with the power of the Stream API, you can write this instead!
+
+```
+double meanMaleHeight = students.toStream()
+                                .filter(student -> student.isMale())
+                                .mapToDouble(student -> student.getHeight())
+                                .average()
+                                .orElse(Double.NaN);
+```
+
+Not only is the numbers of lines reduced (almost by half!), the code utilising streams is also rather intuitive. First, we filter the students who are male, using a lambda expression (`student -> student.isMale()`). Then, we get the heights of these students and followed by the average value of these heights. If the average value does not exist (which can happen when there are no male students), we store NaN in `meanMaleHeight` instead.
+
+Besides its brevity, another cool feature of streams is that code can be executed in parallel using your multicore processor, which leads to faster performance. And the best part is this can be done just by using a simple method call in the Stream API, you don't have implement multithreading or worry about how to go about splitting the work for it to work in parallel.
```

Comment:
Perhaps can give a brief hint on how it executes in parallel -> the students are processed all at once, rather than one-by-one.